GDevelop.js/scripts/generate-dts.mjs

```diff
@@ -0,0 +1,214 @@
+  ['DOMString', 'string'],
+  ['long', 'number'],
+  ['double', 'number'],
+  ['void', 'void'],
+  ['boolean', 'boolean'],
+]);
+
+const none = Symbol('No default value');
+class Parser {
+  static i = 0;
+  /** @private */
+  static str = '';
+  /** @private */
+  static l = 0;
+
+  /** @param {string} str */
+  static set target(str) {
+    this.i = 0;
+    this.str = str;
+    this.l = str.length;
+  }
+  static get done() {
+    return this.i >= this.l;
+  }
+  static get c() {
+    return this.str.charAt(this.i);
+  }
+  static get cc() {
+    return this.str.charCodeAt(this.i);
+  }
+  static skipWhitespaces() {
+    while (this.cc <= 32) this.i++;
+    // Comments go everywhere whitespaces go and must be skipped as if it were whitespace
+    if (this.c === '/') {
+      this.i++;
+      if (this.c === '/') {
+        // //-style comments
+        this.skipUntil('\n');
+      } else if (this.c === '*') {
+        // /* */-style comments
+        do {
+          this.skipUntil('*');
+        } while (this.c !== '/');
+        this.i++;
+      } else console.warn(`Unexpected slash!`);
+      this.skipWhitespaces();
+    }
+  }
+  /** @param {string} thisCharacter */
+  static skipUntil(thisCharacter, skipOverIt = true) {
+    while (this.c !== thisCharacter && !this.done) this.i++;
+    if (skipOverIt) this.i++;
+  }
+  /** @param {string} thisCharacter */
+  static readUntil(thisCharacter, skipOverIt = true) {
+    let token = '';
+    while (this.c !== thisCharacter) {
+      token += this.c;
+      this.i++;
+    }
+    if (skipOverIt) this.i++;
+    return token;
+  }
+
+  static readType() {
+    // Ignore [Const] and such annotations
+    Parser.skipWhitespaces();
+    if (Parser.c === '[') Parser.skipUntil(']');
+    Parser.skipWhitespaces();
+
+    // Read the type
+    let type = Parser.readUntil(' ');
+    let optional = false;
+    if (type === 'optional') optional = true;
+    while (type === 'unsigned' || type === 'optional') {
+      // Re-read the type since unsigned is an unnecessary prefix for typescript
+      Parser.skipWhitespaces();
+      type = Parser.readUntil(' ');
+    }
+    Parser.skipWhitespaces();
+
+    return { type, optional };
+  }
+
+  static readIdentifier() {
+    let name = '';
+    while (
+      (Parser.cc >= 97 && Parser.cc <= 122) || // [a-z]
+      (Parser.cc >= 65 && Parser.cc <= 90) || // [A-Z]
+      (Parser.cc >= 48 && Parser.cc <= 57) || // [1-9]
+      Parser.c === '_'
+    ) {
+      name += Parser.c;
+      Parser.i++;
+    }
+    return name;
+  }
+
+  static readNumber() {
+    let number = '';
+    while (
+      Parser.cc >= 48 &&
+      Parser.cc <= 57 // [1-9]
+    ) {
+      number += Parser.c;
+      Parser.i++;
+    }
+    return parseInt(number, 10);
+  }
+}
+
+const interfaces = [];
+for (const [a, interfaceName, interfaceCode] of bindingsFile.matchAll(
+  /interface ([a-zA-Z]+) {\r?\n?([^}]*)\r?\n}/gm
+)) {
+  const methods = [];
+
+  Parser.target = interfaceCode;
```

Comment:
There is a slight chance the Parser keeps wrongfully keep its state between calls - would have it been clearer/safer to just create a new Parser object? (you don't need to run two Parser at the same time, but that's another reason making it safer compared to something with static variables everywhere. It's also harder to test)

Reply:
Since it is such a small ultra-specific parser only intended for usage within that single simple script I don't think it is necessary. The parser by itself doesn't need to be tested, if anything needs to be tested it is the whole script at once not its components. With the way that code is currently written, there is also not the slightest chance to keep a wrongful state. 

I would agree with this point if we were talking about something that would be reused across other files, or was used in numerous places, or has a lot of changing requirements and mustn't break while enduring numerous modifications, or was a critical part of the software. But since this is just a small script that presumably rarely will be edited, even more rarely in a way that could bring the parser to break, and the parser is only used within a single place.

My reasoning for originally doing this is that recreating a parser for every interface would slow things down and use more memory, which, in the long term, might get more and more annoying as we add new interfaces to GDCore :(